christfellowshipchurch/web-app
# [Chat] Cleanup StreamChatClient initialization and support Guests

# Related Todos
- [**🏕 Clean up Stream client management approach**](https://3.basecamp.com/3926363/buckets/17566548/todos/2982516997)
- [**🏕 Allow guest viewing of chat**](https://3.basecamp.com/3926363/buckets/17566548/todos/2964049257)

# Description
**Main Changes**
- Overhauls `StreamChat` client management logic to be more production-ready, and support all auth state flows
- Allows joining channel as guest when logged out
- **✨New✨** Queries a users' channel role (i.e. "is moderator") so we can use it later

**Tech Changes**
- Adds i18n localization config to `StreamChat`, set to `en` for now
- Moves `src/src/content-single/EventContentItem/components/EventDetails.js` up a level and uses `.sandbox.js` instead, to minimize (my own) confusion having two `EventDetail.js` files only a folder away from each other
- Fixes a React propType warning in `src/ui/Share` component
- **✨New✨** Adds ability to add `?debug` url param to sandbox Events page to show extra 

# Tradeoffs/Caveats
I have not fiddled with guest permissions in Stream yet. Guest users have the ability to send messages by default, but I will remove that ability later.

Nothing is being done with the user role data, and there is a utility and enum definition awkwardly at the top of `EventChat.js`, that I will find a better home for soon. (Or leave it, and remove the code comments 🙃). 

# Test Instructions
1. Make sure the `.env` for **web-app** has `REACT_APP_ENABLE_SANDBOX=true`
1. Start the **web-app** and **apollos-api** locally
1. Navigate to an event page (http://localhost:3000/events/test-24-hour-live-stream)
1. When logged out, verify messages you send are as a guest.
1. Log in without refreshing, verify chat re-initializes and new messages sent are from _you_
1. Refresh the page, so it loads in logged in state. Verify _you_ can send messages.
1. Log out, without refreshing. Verify chat re-initializes as a guest.
1. Open a new Incognito window, to empty all cache and local data, and verify things still work as expected
1. Etc

**Testing Debug info**
1. On an Events page with sandbox enabled, add `?debug` to the end of the url
1. Verify some extra info shows below the message input area in chat  

# Screenshots
<img src="https://user-images.githubusercontent.com/1812955/92194517-e4a7c180-ee38-11ea-9fdc-eaeb5e8a27be.gif" width="100%" />

# Tests
